README.md
# japanese-bookkeeping-svg

[![Build Status](https://travis-ci.org/gunyarakun/japanese-bookkeeping-svg.svg?branch=master)](https://travis-ci.org/gunyarakun/japanese-bookkeeping-svg)
[![Gem Version](https://badge.fury.io/rb/japanese-bookkeeping-svg.svg)](https://badge.fury.io/rb/japanese-bookkeeping-svg)
![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)

Generate SVG files for Japanese style bookkeeping diagrams.

:warning: This SVG uses `textLength` attritube which is not be supported by some SVG viewer. Use [Firefox](https://www.mozilla.org/en-US/firefox/products/), [Batik](https://xmlgraphics.apache.org/batik/), etc.

## Usage

```bash
gem install japanese-bookkeeping-svg
```

仕訳

```ruby
require 'japanese-bookkeeping-svg'

File.open('journal.svg', 'w') do |file|
  file.write JapaneseBookkeepingSVG.journalization(
    debits={
      現金預金: 4000
    },
    credits={
      売上: 2000,
      売掛金: 2000
    }
  ).to_s
end
```

T字勘定

```ruby
require 'japanese-bookkeeping-svg'

File.open('t-accounts.svg', 'w') do |file|
  file.write JapaneseBookkeepingSVG.t_accounts(
    '現金預金',
    debits={
      資本金: 10000
    }
    credits={
      仕入: 2000,
      普通預金: 1234,
      諸口: 3456,
      次月繰越: 3310
    },
  ).to_s
end
```

## Conversion to PNG with Batik

Make `user.css`.

```css
text {
  font-size: 16px;
  font-family: "ヒラギノ明朝 ProN W6", "HiraMinProN-W6";
}
```

Execute. Set `-w 2000` for changing image size to make resolutions up.

```bash
java -jar batik-1.8/batik-rasterizer-1.8.jar -w 2000 -cssUser user.css out.svg
```

## License

``japanese-bookkeeping-svg`` is distributed under the terms of the MIT license (see LICENSE.txt).


ChangeLog
== 2016-06-26 version 0.0.2

* Improve appearances
* Add Batik usage on README.md


== 2016-06-25 version 0.0.1

* First release



test/basic_test.rb
# rubocop:disable Metrics/MethodLength
require 'test_helper'

class BasicTest < Minitest::Test
  def test_journalization
    svg = JapaneseBookkeepingSVG.journalization(
      {
        現金預金: 4000
      },
      {
        売上: 2000,
        売掛金: 2000
      }
    )

    assert_equal <<SVG.chomp, svg.to_s
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="500" height="48" version="1.1" xmlns="http://www.w3.org/2000/svg">
<text x="0" y="16"><tspan x="0" dy="0em">(</tspan></text>
<text x="15" y="16" textLength="110" lengthAdjust="spacing"><tspan x="15" dy="0em">現金預金</tspan></text>
<text x="135" y="16"><tspan x="135" dy="0em">)</tspan></text>
<text x="240" y="16" text-anchor="end"><tspan x="240" dy="0em">4,000</tspan></text>
<text x="260" y="16"><tspan x="260" dy="0em">(</tspan></text>
<text x="275" y="16" textLength="110" lengthAdjust="spacing"><tspan x="275" dy="0em">売上</tspan></text>
<text x="395" y="16"><tspan x="395" dy="0em">)</tspan></text>
<text x="500" y="16" text-anchor="end"><tspan x="500" dy="0em">2,000</tspan></text>
<text x="260" y="32"><tspan x="260" dy="0em">(</tspan></text>
<text x="275" y="32" textLength="110" lengthAdjust="spacing"><tspan x="275" dy="0em">売掛金</tspan></text>
<text x="395" y="32"><tspan x="395" dy="0em">)</tspan></text>
<text x="500" y="32" text-anchor="end"><tspan x="500" dy="0em">2,000</tspan></text>
</svg>
SVG
  end

  def test_t_accounts
    svg = JapaneseBookkeepingSVG.t_accounts(
      '現金預金',
      {
        資本金: 10_000
      },
      {
        仕入: 2000,
        通信費: 500,
        次月繰越: 7500
      }
    )

    assert_equal <<SVG.chomp, svg.to_s
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="500" height="120" version="1.1" xmlns="http://www.w3.org/2000/svg">
<text x="250" y="16" text-anchor="middle" textLength="110" lengthAdjust="spacing"><tspan x="250" dy="0em">現金預金</tspan></text>
<line x1="0" y1="24" x2="500" y2="24" stroke="black" stroke-width="1" />
<text x="15" y="45" textLength="110" lengthAdjust="spacing"><tspan x="15" dy="0em">資本金</tspan></text>
<text x="240" y="45" text-anchor="end"><tspan x="240" dy="0em">10,000</tspan></text>
<text x="265" y="45" textLength="110" lengthAdjust="spacing"><tspan x="265" dy="0em">仕入</tspan></text>
<text x="490" y="45" text-anchor="end"><tspan x="490" dy="0em">2,000</tspan></text>
<text x="265" y="61" textLength="110" lengthAdjust="spacing"><tspan x="265" dy="0em">通信費</tspan></text>
<text x="490" y="61" text-anchor="end"><tspan x="490" dy="0em">500</tspan></text>
<text x="265" y="77" textLength="110" lengthAdjust="spacing"><tspan x="265" dy="0em">次月繰越</tspan></text>
<text x="490" y="77" text-anchor="end"><tspan x="490" dy="0em">7,500</tspan></text>
<line x1="15" y1="85" x2="135" y2="53" stroke="black" stroke-width="1" />
<line x1="15" y1="85" x2="250" y2="85" stroke="black" stroke-width="1" />
<line x1="385" y1="85" x2="500" y2="85" stroke="black" stroke-width="1" />
<text x="240" y="104" text-anchor="end"><tspan x="240" dy="0em">10,000</tspan></text>
<text x="490" y="104" text-anchor="end"><tspan x="490" dy="0em">10,000</tspan></text>
<line x1="135" y1="112" x2="250" y2="112" stroke="black" stroke-width="1" />
<line x1="385" y1="112" x2="500" y2="112" stroke="black" stroke-width="1" />
<line x1="135" y1="115" x2="250" y2="115" stroke="black" stroke-width="1" />
<line x1="385" y1="115" x2="500" y2="115" stroke="black" stroke-width="1" />
<line x1="250" y1="24" x2="250" y2="115" stroke="black" stroke-width="1" />
</svg>
SVG
  end
end


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'japanese-bookkeeping-svg'

require 'minitest/autorun'


lib/japanese-bookkeeping-svg.rb
# rubocop:disable Style/FileName
require_relative 'japanese_bookkeeping_svg/svg_generator'
require_relative 'japanese_bookkeeping_svg/gem_version'

module JapaneseBookkeepingSVG
  WIDTH = 500
  CENTER_X = WIDTH / 2
  LEFT_MARGIN = 15
  RIGHT_MARGIN = 10
  ACCOUNT_WIDTH = 110
  AMOUNT_X = LEFT_MARGIN + ACCOUNT_WIDTH + RIGHT_MARGIN
  AMOUNT_WIDTH = 105

  @@line_height = SVGGenerator.convert_unit('1em') # rubocop:disable Style/ClassVars

  def self.line_height
    @@line_height
  end

  def self.value_style
    {
      'text-anchor' => 'end'
    }
  end

  def self.account_value(svg, x, y, journals, with_brackets) # rubocop:disable Metrics/MethodLength,Metrics/AbcSize
    key_style = {
      'textLength' => ACCOUNT_WIDTH
    }
    sum_value = 0
    journals.map do |key, value|
      c_x = x
      y += line_height
      key_style['lengthAdjust'] = key.size > 8 ? 'spacingAndGlyphs' : 'spacing'
      svg.text(c_x, y, '(') if with_brackets
      c_x += LEFT_MARGIN
      svg.text(c_x, y, key.to_s, key_style)
      c_x += ACCOUNT_WIDTH + RIGHT_MARGIN
      svg.text(c_x, y, ')') if with_brackets
      c_x += AMOUNT_WIDTH
      svg.text(c_x, y, delimited_number(value).to_s, value_style)
      sum_value += value
    end

    [y, sum_value]
  end

  def self.journalization(debits, credits)
    # TODO: check amount
    SVGGenerator.new do
      JapaneseBookkeepingSVG.account_value(self, 0, 0, debits, true)
      JapaneseBookkeepingSVG.account_value(self, CENTER_X + RIGHT_MARGIN, 0, credits, true)
    end
  end

  def self.t_accounts(account, debits, credits) # rubocop:disable Metrics/LineLength,Metrics/MethodLength,Metrics/AbcSize
    # TODO: calc amount
    line_height = SVGGenerator.convert_unit('1em')
    that = self
    SVGGenerator.new do
      account_style = {
        'textLength' => ACCOUNT_WIDTH,
        'text-anchor' => 'middle'
      }
      line_style = {
        'stroke' => 'black',
        'stroke-width' => 1
      }

      account_style['lengthAdjust'] = account.size > 8 ? 'spacingAndGlyphs' : 'spacing'
      text(CENTER_X, line_height, account, account_style)
      t_start_y = line_height + 8
      line(0, t_start_y, WIDTH, t_start_y, line_style)

      y = t_start_y + 5
      debits_y, sum_debits = that.account_value(self, 0, y, debits, false)
      credits_y, sum_credits = that.account_value(self, CENTER_X, y, credits, false)

      if debits_y < credits_y
        y = credits_y + 8
        line(LEFT_MARGIN, y, AMOUNT_X, debits_y + 8, line_style)
        line(LEFT_MARGIN, y, CENTER_X, y, line_style)
        line(CENTER_X + AMOUNT_X, y, WIDTH, y, line_style)
      elsif debits_y > credits_y
        y = debits_y + 8
        line(AMOUNT_X, y, CENTER_X, y, line_style)
        line(CENTER_X + LEFT_MARGIN, y, CENTER_X + AMOUNT_X, credits_y + 8, line_style)
        line(CENTER_X + LEFT_MARGIN, y, WIDTH, y, line_style)
      else # eq
        y = debits_y + 8
        line(AMOUNT_X, y, CENTER_X, y, line_style)
        line(CENTER_X + AMOUNT_X, y, WIDTH, y, line_style)
      end

      y += line_height + 3
      text(CENTER_X - RIGHT_MARGIN, y, that.delimited_number(sum_debits).to_s, that.value_style)
      text(WIDTH - RIGHT_MARGIN, y, that.delimited_number(sum_credits).to_s, that.value_style)

      y += 8
      line(AMOUNT_X, y, CENTER_X, y, line_style)
      line(CENTER_X + AMOUNT_X, y, WIDTH, y, line_style)
      y += 3
      line(AMOUNT_X, y, CENTER_X, y, line_style)
      line(CENTER_X + AMOUNT_X, y, WIDTH, y, line_style)

      line(CENTER_X, t_start_y, CENTER_X, y, line_style)
    end
  end

  def self.delimited_number(number)
    number.to_s.gsub(/(\d)(?=(\d\d\d)+(?!\d))/, '\\1,')
  end
end


lib/japanese_bookkeeping_svg/svg_generator.rb
module JapaneseBookkeepingSVG
  class SVGGenerator # rubocop:disable Metrics/ClassLength
    def initialize(&block)
      @output = []
      @max_width = 0
      @max_height = 0

      return unless block
      instance_exec(&block)
      close
    end

    def close(width = @max_width, height = @max_height)
      unshift_header(width, height)
      push_footer
    end

    def to_s
      @output.join("\n")
    end

    def line(x1, y1, x2, y2, style = {})
      s = style.clone
      attrs = ['stroke', 'stroke-width'].map do |attr|
        if s.key?(attr)
          %( #{attr}="#{s.delete attr}")
        else
          ''
        end
      end.join('')
      @output << %(<line x1="#{x1}" y1="#{y1}" x2="#{x2}" y2="#{y2}"#{attrs}#{style_attr(s)} />)
      update_max_width_and_height(x2, y2)
    end

    def text(x, y, text, style = {}) # rubocop:disable Metrics/MethodLength,Metrics/AbcSize,Metrics/CyclomaticComplexity
      s = style.clone
      attrs = {}
      attrs_str = ['font-family', 'font-size', 'text-anchor', 'textLength', 'lengthAdjust', 'xml:space'].map do |attr|
        if s.key?(attr)
          attrs[attr] = s[attr]
          %( #{attr}="#{s.delete attr}")
        else
          ''
        end
      end.join('')

      text_lines = []
      text_lines << %(<text x="#{x}" y="#{y}"#{attrs_str}#{style_attr(s)}>)
      max_width_em = 0
      height_em = 0
      dy = 0
      text.each_line do |line|
        line.chomp!
        text_lines << %(<tspan x="#{x}" dy="#{dy}em">#{line}</tspan>)
        dy = 1
        max_width_em = line.size if line.size > max_width_em
        height_em += 1
      end
      text_lines << '</text>'
      @output << text_lines.join('')

      max_width = self.class.convert_unit(attrs['textLength'] || "#{max_width_em}em")
      case attrs['text-anchor']
      when nil, 'start'
        max_x = x + max_width
      when 'middle'
        max_x = x + max_width / 2
      when 'end'
        max_x = x
      end
      max_y = y + self.class.convert_unit("#{height_em}em")
      update_max_width_and_height(max_x, max_y)
    end

    def self.convert_unit(length) # rubocop:disable Metrics/AbcSize,Metrics/CyclomaticComplexity,Metrics/MethodLength
      return nil if length.nil?
      m = /\A(\d+)(em|ex|px|in|cm|mm|pt|pc)?\z/.match(length.to_s)
      raise ArgumentError, "Unknown length #{length}" unless m
      v = m[1].to_i
      return v if m[2].nil?
      case m[2].intern
      when :em, :ex
        v * 16
      when :px
        v
      when :in
        v * 96
      when :cm
        (v * 96) / 2.54
      when :mm
        (v * 96) / 25.4
      when :pt
        (v * 4).fdiv(3)
      when :pc
        v * 16
      end
    end

    private

    def update_max_width_and_height(width, height)
      @max_width = [width, @max_width].max
      @max_height = [height, @max_height].max
    end

    def style_attr(style)
      return '' if style.empty?
      ' style="' + style.map do |key, value|
        "#{key}:#{value}"
      end.join(' ') + '"'
    end

    def unshift_header(width, height)
      @output.unshift(<<EOT.chomp)
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg width="#{width}" height="#{height}" version="1.1" xmlns="http://www.w3.org/2000/svg">
EOT
    end

    def push_footer
      @output << '</svg>'
    end
  end
end
